src/blockcrawler/evm/rpc.py

```diff
@@ -195,6 +210,119 @@ async def get_block(self, block_num: HexInt, full_transactions: bool = False) ->
         )
         return block
 
+    async def get_block_by_timestamp(self, blockchain: BlockChain, timestamp: HexInt) -> EvmBlock:
+        """Get block number by timestamp.
+
+        :param blockchain: Blockchain name as a reference on when did the
+            blockchain starts doing Point-of-Stake.
+        :param timestamp: The timestamp of the block you wish to get.
+        :returns: The block which timestamp is equal to given timestamp
+            or the nearest block if there's no block with the exact timestamp.
+        """
+
+        return await self.__get_block_by_timestamp(blockchain, timestamp)
+
+    async def __get_block_by_timestamp(
+        self,
+        blockchain: BlockChain,
+        timestamp: HexInt,
+        left_block: Optional[EvmBlock] = None,
+        right_block: Optional[EvmBlock] = None,
+    ) -> EvmBlock:
+        """Recursively get block number by timestamp until exact or nearest match.
+
+        :param blockchain: Blockchain name as a reference on when did the
+            blockchain starts doing Point-of-Stake.
+        :param timestamp: The timestamp of the block you wish to get.
+        :param left_block: The left bound block limit of binary search.
+        :param right_block: The right bound block limit of binary search.
+        :returns: The block which timestamp is equal to given timestamp
+            or the nearest block if there's no block with the exact timestamp.
+        """
+
+        # Get block by timestamp binary search source: https://ethereum.stackexchange.com/a/127720
+
+        if not left_block or not right_block:  # no recursions made yet
+            pos_info: dict = self.__POS_START_TIMESTAMP_BLOCK_MAP.get(blockchain, {})
+            pos_timestamp: Optional[HexInt] = pos_info.get("timestamp")
+            pos_block_number: Optional[HexInt] = pos_info.get("block_number")
+
+            # Check whether timestamp is a known value
+            if pos_info and timestamp == pos_timestamp:
+                return await self.get_block(pos_block_number)  # type: ignore
+
+            right_block = await self.get_block(await self.get_block_number())
```

Comment:
@aenglander 
I'm frequently getting `blockcrawler.core.rpc.RpcClientError: Error retrieving block: no block returned` when I want to get latest block by that code. Any suggestion?

Reply:
I'm planning to create new PR for this one as `EvmRpcClient.get_latest_block`.